.github/actions/gradle/experiment-1/action.yml

```
@@ -110,25 +113,36 @@ runs:
           ${ARG_DEVELOCITY_URL:+"-s" "$ARG_DEVELOCITY_URL"} \
           ${ARG_DEVELOCITY_ENABLE:+"-e"} \
           ${RUNNER_DEBUG:+"--debug"}
+        EXPERIMENT_EXIT_CODE=$?
 
         # Set the Build Scan urls as outputs
-        RECEIPT_FILE=".data/01-validate-incremental-building/latest/exp1-*.receipt"
         BUILD_SCAN_1=$(grep -m 1 "first build" ${RECEIPT_FILE} | sed 's/.* //')
         BUILD_SCAN_2=$(grep -m 1 "second build" ${RECEIPT_FILE} | sed 's/.* //')
         
         echo "buildScanFirstBuild=$BUILD_SCAN_1" >> $GITHUB_OUTPUT
         echo "buildScanSecondBuild=$BUILD_SCAN_2" >> $GITHUB_OUTPUT
         
-        cat $RECEIPT_FILE >> $GITHUB_STEP_SUMMARY
+        # Copy receipt to a predictable location
+        RECEIPT_FILE=".data/01-validate-incremental-building/latest/exp1-*.receipt"
+        if [ -f ${RECEIPT_FILE} ]; then
+          cp ${RECEIPT_FILE} ../receipt.txt
+        fi
```

What do you think about something like this?

```suggestion
        echo "receiptFile=$(readlink -f .data/01-validate-incremental-building/latest/exp1-*.receipt)" >> $GITHUB_OUTPUT
```

That way we don't have to copy files around to be used in a later step. I think that would also allow users to consume it and could write the file contents to Slack or an email or something without having to worry about where the file is located. I think that's true anyway.

Then later we can do:

```shell
if [ -f "${{ steps.run.outputs.receiptFile }}" ]; then
...
fi
```

---

To expose the file outside of the composite action, we would also need to declare it as an extra output [there](https://github.com/gradle/develocity-build-validation-scripts/blob/e173ee8cf2815af3b687701b566c3ce39f04653f/.github/actions/gradle/experiment-1/action.yml#L40).
We could do that but my intention is exactly opposite as described there
> The receipt file is renamed to a predictable receipt.txt file to avoid passing the filename across steps and adding an attack surface as the file content is archived and inlined in the GitHub summary

We also could keep the reference `.data/01-validate-incremental-building/latest/exp1-*.receipt` in the 3 different steps to avoid the `copy`
The `copy` is just there to avoid repeating the uneasy reference.
Do you think it would be better?

---

> To expose the file outside of the composite action, we would also need to declare it as an extra output [there](https://github.com/gradle/develocity-build-validation-scripts/blob/e173ee8cf2815af3b687701b566c3ce39f04653f/.github/actions/gradle/experiment-1/action.yml#L40).

Right, I knew there was something I was missing. I would do that.

> adding an attack surface as the file content is archived and inlined in the GitHub summary

I don't get this part.

---

Having as a step input a file path which you then display in the summary and upload as an artifact is something I am a bit reluctant to do

---

Sorry, but I still don't understand. We already displayed the content in the summary before here.

https://github.com/gradle/develocity-build-validation-scripts/blob/159dc34d6de1b2019eb9f9807d2ff14a2b90ff1c/.github/actions/gradle/experiment-1/action.yml#L122

And the file was an input to the upload action here.

https://github.com/gradle/develocity-build-validation-scripts/blob/159dc34d6de1b2019eb9f9807d2ff14a2b90ff1c/.github/actions/gradle/experiment-1/action.yml#L129

---

in both cases, the file name was configured explicitly to be `develocity-gradle-build-validation/.data/01-validate-incremental-building/latest*/exp1-*.receipt `
I don't want to create a step leaking the content of a file which is defined as a previous step output.
I can imagine a scenario where this output could be overriden with a path to a file containing secrets in the consuming workflow.
You can argue that there are other easy ways to capture the file in such conditions, but I'd rather not create a step like this

---

The artifact is uploaded immediately after the experiment completes in the following step and the summary is written immediately after that. All of this happens within the composite action so the user can overwrite the file, but neither the uploaded artifact or experiment summary should be affected. 

In any case, they could have overwritten the file before these changes too.

---

I agree that the risk is limited or not worsened but I still feel it is bad practice to display a file content without controlling how the file path is obtained

---

Is your concern only about exposing the file path to the user, i.e., adding it as an output of the composite action?

---

My concern is that the composite step is displaying a file which could potentially be something else than the report.
You could think of a scenario with a complex consuming workflow, calling this composite, with maintainers thoroughly checking that files with secrets are not undisclosed by a PR, but omitting to control the behavior of steps brought by composite actions.
This is unlikely to happen but not impossible if we make the receipt file a step input/output

---

We agreed on a call to inline the full receipt path when used